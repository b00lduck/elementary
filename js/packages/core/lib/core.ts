import {
  createNode,
  isNode,
  resolve,
  ElemNode,
  NodeRepr_t,
} from "../nodeUtils";

export function constant(props: { key?: string; value: number }): NodeRepr_t {
  return createNode("const", props, []);
}

export function sr(): NodeRepr_t {
  return createNode("sr", {}, []);
}

export function time(): NodeRepr_t {
  return createNode("time", {}, []);
}

export function counter(gate: ElemNode): NodeRepr_t {
  return createNode("counter", {}, [resolve(gate)]);
}

export function accum(xn: ElemNode, reset: ElemNode): NodeRepr_t {
  return createNode("accum", {}, [resolve(xn), resolve(reset)]);
}

// Phasor node
export function phasor(rate: ElemNode): NodeRepr_t {
  return createNode("phasor", {}, [resolve(rate)]);
}

export function syncphasor(rate: ElemNode, reset: ElemNode): NodeRepr_t {
  return createNode("sphasor", {}, [resolve(rate), resolve(reset)]);
}

export function latch(t: ElemNode, x: ElemNode): NodeRepr_t {
  return createNode("latch", {}, [resolve(t), resolve(x)]);
}

export function maxhold(
  props: { key?: string; hold?: number },
  x: ElemNode,
  reset: ElemNode,
): NodeRepr_t {
  return createNode("maxhold", props, [resolve(x), resolve(reset)]);
}

export function once(
  props: { key?: string; arm?: boolean },
  x: ElemNode,
): NodeRepr_t {
  return createNode("once", props, [resolve(x)]);
}

export function rand(props?: { key?: string, seed?: number }): NodeRepr_t {
  return createNode("rand", props || {}, []);
}

export function metro(props?: {
  key?: string;
  name?: string;
  interval?: number;
}): NodeRepr_t {
  return createNode("metro", props || {}, []);
}

export function sample(
  props: {
    key?: string;
    path: string;
    mode?: string;
    startOffset?: number;
    stopOffset?: number;
  },
  trigger: ElemNode,
  rate: ElemNode,
): NodeRepr_t {
  return createNode("sample", props, [resolve(trigger), resolve(rate)]);
}

export function table(
  props: {
    key?: string;
    path: string;
  },
  t: ElemNode,
): NodeRepr_t {
  return createNode("table", props, [resolve(t)]);
}

export function convolve(
  props: {
    key?: string;
    path: string;
  },
  x: ElemNode,
): NodeRepr_t {
  return createNode("convolve", props, [resolve(x)]);
}

export function seq(
  props: {
    key?: string;
    seq: Array<number>;
    offset?: number;
    hold?: boolean;
    loop?: boolean;
  },
  trigger: ElemNode,
  reset: ElemNode,
): NodeRepr_t {
  return createNode("seq", props, [resolve(trigger), resolve(reset)]);
}

export function seq2(
  props: {
    key?: string;
    seq: Array<number>;
    offset?: number;
    hold?: boolean;
    loop?: boolean;
  },
  trigger: ElemNode,
  reset: ElemNode,
): NodeRepr_t {
  return createNode("seq2", props, [resolve(trigger), resolve(reset)]);
}

export function sparseq(
  props: {
    key?: string;
    seq: Array<{ value: number; tickTime: number }>;
    offset?: number;
    loop?: boolean | Array<number>;
    interpolate?: number;
    tickInterval?: number;
  },
  trigger: ElemNode,
  reset: ElemNode,
): NodeRepr_t {
  return createNode("sparseq", props, [resolve(trigger), resolve(reset)]);
}

export function sparseq2(
  props: {
    key?: string;
    seq: Array<{ value: number; time: number }>;
  },
  time: ElemNode,
): NodeRepr_t {
  return createNode("sparseq2", props, [resolve(time)]);
}

export function sampleseq(
  props: {
    key?: string;
    path: string;
    seq: Array<{ value: number; time: number }>;
    duration: number;
  },
  time: ElemNode,
): NodeRepr_t {
  return createNode("sampleseq", props, [resolve(time)]);
}

export function sampleseq2(
  props: {
    key?: string;
    path: string;
    seq: Array<{ value: number; time: number }>;
    duration: number;
    stretch?: number;
    shift?: number;
  },
  time: ElemNode,
): NodeRepr_t {
  return createNode("sampleseq2", props, [resolve(time)]);
}

export function pole(p: ElemNode, x: ElemNode): NodeRepr_t {
  return createNode("pole", {}, [resolve(p), resolve(x)]);
}

export function env(
  atkPole: ElemNode,
  relPole: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("env", {}, [
    resolve(atkPole),
    resolve(relPole),
    resolve(x),
  ]);
}

// Single sample delay node
export function z(x: ElemNode): NodeRepr_t {
  return createNode("z", {}, [resolve(x)]);
}

// Variable delay node
export function delay(
  props: {
    key?: string;
    size: number;
  },
  len: ElemNode,
  fb: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("delay", props, [resolve(len), resolve(fb), resolve(x)]);
}

// Static delay node
export function sdelay(
  props: {
    key?: string;
    size: number;
  },
  x: ElemNode,
): NodeRepr_t {
  return createNode("sdelay", props, [resolve(x)]);
}

// Multimode1p
export function prewarp(fc: ElemNode): NodeRepr_t {
  return createNode("prewarp", {}, [resolve(fc)]);
}

export function mm1p(
  props: {
    key?: string;
    mode?: string;
  },
  fc: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("mm1p", props, [resolve(fc), resolve(x)]);
}

export function svf(
  props: {
    key?: string;
    mode?: string;
  },
  fc: ElemNode,
  q: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("svf", props, [resolve(fc), resolve(q), resolve(x)]);
}

export function svfshelf(
  props: {
    key?: string;
    mode?: string;
  },
  fc: ElemNode,
  q: ElemNode,
  gainDecibels: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("svfshelf", props, [
    resolve(fc),
    resolve(q),
    resolve(gainDecibels),
    resolve(x),
  ]);
}

export function biquad(
  b0: ElemNode,
  b1: ElemNode,
  b2: ElemNode,
  a1: ElemNode,
  a2: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("biquad", {}, [
    resolve(b0),
    resolve(b1),
    resolve(b2),
    resolve(a1),
    resolve(a2),
    resolve(x),
  ]);
}

// Feedback nodes
export function tapIn(props: { key?: string; name: string }): NodeRepr_t {
  return createNode("tapIn", props, []);
}

export function tapOut(
  props: { key?: string; name: string },
  x: ElemNode,
): NodeRepr_t {
  return createNode("tapOut", props, [resolve(x)]);
}

// Analysis
export function meter(
  props: {
    key?: string;
    name?: string;
  },
  x: ElemNode,
): NodeRepr_t {
  return createNode("meter", props, [resolve(x)]);
}

export function snapshot(
  props: {
    key?: string;
    name?: string;
  },
  trigger: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("snapshot", props, [resolve(trigger), resolve(x)]);
}

export function scope(
  props: {
    key?: string;
    name?: string;
    size?: number;
    channels?: number;
  },
  ...args: Array<ElemNode>
): NodeRepr_t {
  return createNode("scope", props, args.map(resolve));
}

export function fft(
  props: {
    key?: string;
    name?: string;
    size?: number;
  },
  x: ElemNode,
): NodeRepr_t {
  return createNode("fft", props, [resolve(x)]);
}

export function capture(
  props: {
    key?: string;
  },
  g: ElemNode,
  x: ElemNode,
): NodeRepr_t {
  return createNode("capture", props, [resolve(g), resolve(x)]);
}
