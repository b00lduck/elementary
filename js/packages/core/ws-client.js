import {el, Renderer, formatRoots} from "./dist/index.js";

// Create WebSocket connection.
const socket = new WebSocket("ws://localhost:8080");

// Connection opened
socket.addEventListener("open", (event) => {
  // Once the connection is opened we generate a new graph every 500ms
  setInterval(() => {
    let i = 1 + Math.round((Math.random() * 8));
    let f = 110 * i;

    socket.send(JSON.stringify({
      graph: formatRoots([el.cycle(el.const({key: 'a', value: f})), el.cycle(el.meter({}, el.const({key: 'b', value: f * 1.01})))]),
    }));
  }, 500);
});

// Listen for messages
socket.addEventListener("message", (event) => {
  console.log("Message from server ", event.data);
});
