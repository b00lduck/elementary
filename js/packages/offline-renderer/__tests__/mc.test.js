import OfflineRenderer from "..";
import { el } from "@elemaudio/core";

test("mc table", async function () {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 1,
    virtualFileSystem: {
      "/v/ones": Float32Array.from([1, 1, 1]),
      "/v/stereo": [
        Float32Array.from([27, 27, 27]),
        Float32Array.from([15, 15, 15]),
      ],
    },
  });

  // Graph
  await core.render(
    el.add(...el.mc.table({ path: "/v/stereo", channels: 2 }, 0)),
  );

  // Ten blocks of data
  let inps = [];
  let outs = [new Float32Array(512 * 10)];

  // Get past the fade-in
  core.process(inps, outs);

  // Process another small block
  inps = [];
  outs = [new Float32Array(16)];

  core.process(inps, outs);

  expect(outs[0]).toMatchSnapshot();

  // Let's try a graph unpacking more channels than the resource has
  await core.render(
    el.add(...el.mc.table({ path: "/v/stereo", channels: 3 }, 0)),
  );

  // Get past the fade-in
  for (let i = 0; i < 100; ++i) {
    core.process(inps, outs);
  }

  // Process another small block
  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();

  // And again unpacking fewer channels
  await core.render(
    el.add(...el.mc.table({ path: "/v/stereo", channels: 1 }, 0)),
  );

  // Get past the fade-in
  for (let i = 0; i < 100; ++i) {
    core.process(inps, outs);
  }

  // Process another small block
  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();

  // Now here we expect that the first graph has been totally nudged out
  // of relevance, except for the nodes that remain shared which are the table
  // and the const 0. If we gc then, we should remove that add() and root() from
  // the original graph.
  expect(await core.gc()).toEqual([1611541315, 1811703364]);

  // But now the act of removing that add node should have pruned the outlet connections
  // list on the table node. We'll test that original graph again to ensure that after
  // rebuilding that connection everything still looks good.
  await core.render(
    el.add(...el.mc.table({ path: "/v/stereo", channels: 2 }, 0)),
  );

  // Get past the fade-in
  for (let i = 0; i < 100; ++i) {
    core.process(inps, outs);
  }

  // Process another small block
  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();
});

test("mc sampleseq", async function () {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 1,
    virtualFileSystem: {
      "/v/stereo": [
        Float32Array.from(Array.from({ length: 128 }).fill(27)),
        Float32Array.from(Array.from({ length: 128 }).fill(15)),
      ],
    },
  });

  let [time, setTimeProps] = core.createRef("const", { value: 0 }, []);

  core.render(
    el.add(
      ...el.mc.sampleseq(
        {
          path: "/v/stereo",
          channels: 2,
          duration: 128,
          seq: [
            { time: 0, value: 0 },
            { time: 128, value: 1 },
            { time: 256, value: 0 },
            { time: 512, value: 1 },
          ],
        },
        time,
      ),
    ),
  );

  // Ten blocks of data to get past the root node fade
  let inps = [];
  let outs = [new Float32Array(10 * 512)];

  // Get past the fade-in
  core.process(inps, outs);

  // Now we expect to see zeros because we're fixed at time 0
  outs = [new Float32Array(32)];
  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();

  // Jump in time
  setTimeProps({ value: 520 });

  // Spin for a few blocks and we should see the gain fade resolve and
  // emit the constant sum of the two channels
  for (let i = 0; i < 10; ++i) {
    core.process(inps, outs);
  }

  expect(outs[0]).toMatchSnapshot();
});

test("mc capture", async function () {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 4,
    blockSize: 32,
  });

  let [gate, setGateProps] = core.createRef("const", { value: 0 }, []);

  core.render(
    ...el.mc.capture({ name: "test", channels: 4 }, gate, 1, 2, 3, 4),
  );

  // Ten blocks of data to get past the root node fade
  let inps = [];
  let outs = [
    new Float32Array(32),
    new Float32Array(32),
    new Float32Array(32),
    new Float32Array(32),
  ];

  // Get past the fade-in
  for (let i = 0; i < 1000; ++i) {
    core.process(inps, outs);
  }

  let eventCallback = jest.fn();
  core.on("mc.capture", eventCallback);

  setGateProps({ value: 1 });
  core.process(inps, outs);
  expect(outs).toMatchSnapshot();

  setGateProps({ value: 0 });
  core.process(inps, outs);

  expect(eventCallback.mock.calls).toHaveLength(1);
  let args = eventCallback.mock.calls[0];
  let evt = args[0];

  expect(evt.data).toHaveLength(4);
  expect(evt.source).toBe("test");

  for (let i = 0; i < 4; ++i) {
    expect(evt.data[i]).toHaveLength(32);

    for (let j = 0; j < 32; ++j) {
      expect(evt.data[i][j]).toBe(i + 1);
    }
  }
});

test("mc.sample", async function () {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 1,
    numOutputChannels: 1,
    blockSize: 32,
    virtualFileSystem: {
      "/v/stereo": [
        Float32Array.from(Array.from({ length: 512 }).fill(27)),
        Float32Array.from(Array.from({ length: 512 }).fill(15)),
      ],
    },
  });

  let [gate, setGateProps] = core.createRef("const", { value: 0 }, []);

  await core.render(
    el.add(
      ...el.mc.sample(
        {
          path: "/v/stereo",
          channels: 2,
        },
        gate,
      ),
    ),
  );

  let inps = [new Float32Array(32)];
  let outs = [new Float32Array(32)];

  // Get past the fade-in
  for (let i = 0; i < 1000; ++i) {
    core.process(inps, outs);
  }

  // Now we expect to see zeros because we haven't triggered the sample
  core.process(inps, outs);
  expect(outs).toMatchObject([new Float32Array(32)]);

  // Trigger and we should see sample playback
  await setGateProps({ value: 1 });

  // Gain fade on sample boundary
  for (let i = 0; i < 5; ++i) {
    core.process(inps, outs);

    for (let j = 1; j < 32; ++j) {
      expect(outs[0][j]).toBeGreaterThanOrEqual(0);
      expect(outs[0][j]).toBeLessThan(42);
      expect(outs[0][j]).toBeGreaterThan(outs[0][j - 1]);
    }
  }

  // Gain fade resolves on the next block
  core.process(inps, outs);

  // Then in the subsequent block we should see constant 42s
  core.process(inps, outs);
  expect(outs).toMatchObject([Float32Array.from({ length: 32 }, () => 42)]);

  // Test for triggers mid-block
  await core.render(
    el.add(
      ...el.mc.sample(
        {
          path: "/v/stereo",
          channels: 2,
          mode: "gate",
        },
        el.in({ channel: 0 }),
      ),
    ),
  );

  // Get past the fade-in
  for (let i = 0; i < 1000; ++i) {
    core.process(inps, outs);
  }

  // Now we send a very intentional trigger
  for (let i = 8; i < 16; ++i) {
    inps[0][i] = 1.0;
  }

  core.process(inps, outs);

  // From 0 to 8 we expect silence
  expect(outs[0].slice(0, 8)).toMatchObject(new Float32Array(8));

  // From 8 to 16 we expect to see the gain ramp up
  for (let i = 8; i < 16; ++i) {
    expect(outs[0][i]).toBeGreaterThanOrEqual(0);
    expect(outs[0][i]).toBeLessThan(42);
    expect(outs[0][i]).toBeGreaterThanOrEqual(outs[0][i - 1]);
  }

  // Then we expect to see the gain ramp down
  for (let i = 17; i < 24; ++i) {
    expect(outs[0][i]).toBeLessThanOrEqual(outs[0][i - 1]);
  }
});

test("mc.sample again", async function () {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 0,
    numOutputChannels: 1,
    blockSize: 32,
    virtualFileSystem: {
      "/v/stereo": [
        Float32Array.from([1, 2, 3, 4, 5, 6, 7, 8]),
        Float32Array.from([1, 2, 3, 4, 5, 6, 7, 8]),
      ],
    },
  });

  await core.render(
    el.add(
      ...el.mc.sample(
        {
          path: "/v/stereo",
          channels: 2,
          mode: "loop",
          key: "test",
        },
        1,
      ),
    ),
  );

  let inps = [];
  let outs = [new Float32Array(32)];

  // Fade in
  for (let i = 0; i < 1000; ++i) {
    core.process(inps, outs);
  }

  expect(outs[0]).toMatchSnapshot();

  await core.render(
    el.add(
      ...el.mc.sample(
        {
          path: "/v/stereo",
          channels: 2,
          mode: "loop",
          key: "test",
          playbackRate: 2.0,
        },
        1,
      ),
    ),
  );

  core.process(inps, outs);
  expect(outs[0]).toMatchSnapshot();

  await core.render(
    el.add(
      ...el.mc.sample(
        {
          path: "/v/stereo",
          channels: 2,
          mode: "loop",
          // We need a new trigger for the start/stop offset to take hold
          key: "test2",
          playbackRate: 2.0,
          startOffset: 1,
          stopOffset: 1,
        },
        1,
      ),
    ),
  );

  // Fade in
  for (let i = 0; i < 1000; ++i) {
    core.process(inps, outs);
  }

  expect(outs[0]).toMatchSnapshot();
});
