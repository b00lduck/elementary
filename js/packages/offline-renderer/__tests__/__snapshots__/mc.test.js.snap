// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mc capture 1`] = `
Array [
  Float32Array [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  Float32Array [
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
  ],
  Float32Array [
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
  ],
  Float32Array [
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
  ],
]
`;

exports[`mc sampleseq 1`] = `
Float32Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`mc sampleseq 2`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;

exports[`mc table 1`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;

exports[`mc table 2`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;

exports[`mc table 3`] = `
Float32Array [
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
  27,
]
`;

exports[`mc table 4`] = `
Float32Array [
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
  42,
]
`;
